hal/src/rtl872x/ble_hal.cpp

```diff
@@ -2147,6 +2229,12 @@ void BleGap::handleDevStateChanged(T_GAP_DEV_STATE newState, uint16_t cause) {
         }
     }
     state_.raw = nState.raw;
+    // In case of race condition in stack
+    RtlGapDevState s;
+    if (le_get_gap_param(GAP_PARAM_DEV_STATE, &s.state) == GAP_CAUSE_SUCCESS && nState.raw != s.raw) {
+        LOCAL_DEBUG("!!!!!!!!! Race condition. Notified: %02X, Get: %02X", nState.raw, s.raw);
+        state_.raw = s.raw;
```

Comment:
My understanding of this function is that we get a state change notification from the BLE Event Dispatcher, which includes the new BLE state. Is this correct?

This new behavior here is to also query the BLE stack for the actual state. If the queried state differs from the notification state, something is out of sync. Its possible the BLE stack changed states again between the original notification and now. Is this correct?

If we are also updating `state_` with the newly queried value, should we also act on the potential changes like we do above? IE check `newState.gap_adv_state != state_.state.gap_adv_state`  etc.

Reply:
#1  Correct 
#2 I don't think the BLE stack state is changed in between. It's most likely that, for example, normally the advertising state changes in the sequence of `GAP_ADV_STATE_ADVERTISING -> GAP_ADV_STATE_STOP -> GAP_ADV_STATE_IDLE`, but sometimes the events are notified in wrong sequence `GAP_ADV_STATE_ADVERTISING -> GAP_ADV_STATE_IDLE -> GAP_ADV_STATE_STOP`. The double check is just to correct the internal cached `state_`.

And yes, we should first check the polled state before the check routine above.